server: TestClusterVersionUpgrade failed

server.TestClusterVersionUpgrade [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4072622&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4072622&tab=artifacts#/) on master @ [d66c14663e027a5d9699b8a15e3d6f5821877ef5](https://github.com/cockroachdb/cockroach/commits/d66c14663e027a5d9699b8a15e3d6f5821877ef5):


```
=== RUN   TestClusterVersionUpgrade
    test_log_scope.go:79: test logs captured to: /artifacts/tmp/_tmp/ab5462d2cb3989e2a33018db970f23de/logTestClusterVersionUpgrade615506918
    test_log_scope.go:80: use -show-logs to present logs inline
    version_cluster_test.go:273: condition failed to evaluate within 45s: 0: v21.2-36 active=true (wanted false)
        goroutine 32856908 [running]:
        runtime/debug.Stack()
        	GOROOT/src/runtime/debug/stack.go:24 +0x65
        github.com/cockroachdb/cockroach/pkg/testutils.SucceedsWithin({0x4e7c7b8, 0xc01357e340}, 0x0, 0x3bb2040)
        	github.com/cockroachdb/cockroach/pkg/testutils/soon.go:60 +0x5f
        github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon({0x4e7c7b8, 0xc01357e340}, 0xc0105ad288)
        	github.com/cockroachdb/cockroach/pkg/testutils/soon.go:41 +0x4a
        github.com/cockroachdb/cockroach/pkg/server_test.TestClusterVersionUpgrade(0xc01357e340)
        	github.com/cockroachdb/cockroach/pkg/server_test/version_cluster_test.go:273 +0x8f6
        testing.tRunner(0xc01357e340, 0x3f84298)
        	GOROOT/src/testing/testing.go:1259 +0x102
        created by testing.(*T).Run
        	GOROOT/src/testing/testing.go:1306 +0x35a
    panic.go:642: -- test log scope end --
--- FAIL: TestClusterVersionUpgrade (48.59s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- TAGS=bazel,gss,deadlock
</p>
</details>
/cc @cockroachdb/server
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestClusterVersionUpgrade.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-12163

server.TestClusterVersionUpgrade [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4132601&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4132601&tab=artifacts#/) on master @ [5ad21e3896ee809e9c3ebc28bb22166f1275acca](https://github.com/cockroachdb/cockroach/commits/5ad21e3896ee809e9c3ebc28bb22166f1275acca):


```
=== RUN   TestClusterVersionUpgrade
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestClusterVersionUpgrade192890838
    test_log_scope.go:80: use -show-logs to present logs inline
    version_cluster_test.go:273: condition failed to evaluate within 3m45s: 0: v21.2-36 active=true (wanted false)
        goroutine 679043 [running]:
        runtime/debug.Stack()
        	/usr/local/go/src/runtime/debug/stack.go:24 +0x72
        github.com/cockroachdb/cockroach/pkg/testutils.SucceedsWithin({0x8aac8d8, 0xc00682e4e0}, 0x89fe620, 0x5a0dfc0)
        	/go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:60 +0x7d
        github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon({0x8aac8d8, 0xc00682e4e0}, 0xc006278d8c)
        	/go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:41 +0x5b
        github.com/cockroachdb/cockroach/pkg/server_test.TestClusterVersionUpgrade(0xc00682e4e0)
        	/go/src/github.com/cockroachdb/cockroach/pkg/server/version_cluster_test.go:273 +0xae6
        testing.tRunner(0xc00682e4e0, 0x5de7e58)
        	/usr/local/go/src/testing/testing.go:1259 +0x230
        created by testing.(*T).Run
        	/usr/local/go/src/testing/testing.go:1306 +0x727
    panic.go:661: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestClusterVersionUpgrade192890838
--- FAIL: TestClusterVersionUpgrade (234.85s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- GOFLAGS=-race -parallel=4
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestClusterVersionUpgrade.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

server.TestClusterVersionUpgrade [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4150627&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4150627&tab=artifacts#/) on master @ [912964e02ddd951c77d4f71981ae18b3894e9084](https://github.com/cockroachdb/cockroach/commits/912964e02ddd951c77d4f71981ae18b3894e9084):


```
=== RUN   TestClusterVersionUpgrade
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestClusterVersionUpgrade2151831843
    test_log_scope.go:80: use -show-logs to present logs inline
    version_cluster_test.go:273: condition failed to evaluate within 45s: 0: v21.2-42 active=true (wanted false)
        goroutine 39263657 [running]:
        runtime/debug.Stack()
        	/usr/local/go/src/runtime/debug/stack.go:24 +0x65
        github.com/cockroachdb/cockroach/pkg/testutils.SucceedsWithin({0x4fbb778, 0xc00866eea0}, 0x0, 0x3cc1280)
        	/go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:60 +0x5f
        github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon({0x4fbb778, 0xc00866eea0}, 0xc00e9845b8)
        	/go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:41 +0x4a
        github.com/cockroachdb/cockroach/pkg/server_test.TestClusterVersionUpgrade(0xc00866eea0)
        	/go/src/github.com/cockroachdb/cockroach/pkg/server/version_cluster_test.go:273 +0x8f6
        testing.tRunner(0xc00866eea0, 0x409b1a0)
        	/usr/local/go/src/testing/testing.go:1259 +0x102
        created by testing.(*T).Run
        	/usr/local/go/src/testing/testing.go:1306 +0x35a
    panic.go:642: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestClusterVersionUpgrade2151831843
--- FAIL: TestClusterVersionUpgrade (49.71s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- TAGS=deadlock

- GOFLAGS=-parallel=4
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestClusterVersionUpgrade.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

server.TestClusterVersionUpgrade [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4152284&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4152284&tab=artifacts#/) on master @ [d1eafdf56b42ca09a169c199916917a144c2c412](https://github.com/cockroachdb/cockroach/commits/d1eafdf56b42ca09a169c199916917a144c2c412):


```
=== RUN   TestClusterVersionUpgrade
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestClusterVersionUpgrade1710923491
    test_log_scope.go:80: use -show-logs to present logs inline
    version_cluster_test.go:273: condition failed to evaluate within 3m45s: 0: v21.2-42 active=true (wanted false)
        goroutine 695653 [running]:
        runtime/debug.Stack()
        	/usr/local/go/src/runtime/debug/stack.go:24 +0x72
        github.com/cockroachdb/cockroach/pkg/testutils.SucceedsWithin({0x8adf5b8, 0xc008edc4e0}, 0x8a30b60, 0x5a36da0)
        	/go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:60 +0x7d
        github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon({0x8adf5b8, 0xc008edc4e0}, 0xc00249c7bc)
        	/go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:41 +0x5b
        github.com/cockroachdb/cockroach/pkg/server_test.TestClusterVersionUpgrade(0xc008edc4e0)
        	/go/src/github.com/cockroachdb/cockroach/pkg/server/version_cluster_test.go:273 +0xae6
        testing.tRunner(0xc008edc4e0, 0x5e12d10)
        	/usr/local/go/src/testing/testing.go:1259 +0x230
        created by testing.(*T).Run
        	/usr/local/go/src/testing/testing.go:1306 +0x727
    panic.go:661: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestClusterVersionUpgrade1710923491
--- FAIL: TestClusterVersionUpgrade (241.98s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- GOFLAGS=-json
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestClusterVersionUpgrade.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

server.TestClusterVersionUpgrade [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4190642&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4190642&tab=artifacts#/) on master @ [8cd28089c6c7333615ba3201e841839001d2f0e1](https://github.com/cockroachdb/cockroach/commits/8cd28089c6c7333615ba3201e841839001d2f0e1):


```
=== RUN   TestClusterVersionUpgrade
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestClusterVersionUpgrade2677744765
    test_log_scope.go:80: use -show-logs to present logs inline
    version_cluster_test.go:273: condition failed to evaluate within 45s: 0: v21.2-48 active=true (wanted false)
        goroutine 42023011 [running]:
        runtime/debug.Stack()
        	/usr/local/go/src/runtime/debug/stack.go:24 +0x65
        github.com/cockroachdb/cockroach/pkg/testutils.SucceedsWithin({0x50063f8, 0xc005dc8000}, 0x0, 0x3cedf40)
        	/go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:60 +0x5f
        github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon({0x50063f8, 0xc005dc8000}, 0xc006c725d0)
        	/go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:41 +0x4a
        github.com/cockroachdb/cockroach/pkg/server_test.TestClusterVersionUpgrade(0xc005dc8000)
        	/go/src/github.com/cockroachdb/cockroach/pkg/server/version_cluster_test.go:273 +0x8f6
        testing.tRunner(0xc005dc8000, 0x40cdba0)
        	/usr/local/go/src/testing/testing.go:1259 +0x102
        created by testing.(*T).Run
        	/usr/local/go/src/testing/testing.go:1306 +0x35a
    panic.go:642: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestClusterVersionUpgrade2677744765
--- FAIL: TestClusterVersionUpgrade (50.12s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- TAGS=deadlock

- GOFLAGS=-parallel=4
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestClusterVersionUpgrade.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

server.TestClusterVersionUpgrade [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4222432&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4222432&tab=artifacts#/) on master @ [ce6d32ad93283556cb88bb69eae595c4febaa6c4](https://github.com/cockroachdb/cockroach/commits/ce6d32ad93283556cb88bb69eae595c4febaa6c4):


```
=== RUN   TestClusterVersionUpgrade
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestClusterVersionUpgrade338561729
    test_log_scope.go:80: use -show-logs to present logs inline
    version_cluster_test.go:273: condition failed to evaluate within 3m45s: 0: v21.2-50 active=true (wanted false)
        goroutine 529114 [running]:
        runtime/debug.Stack()
        	/usr/local/go/src/runtime/debug/stack.go:24 +0x72
        github.com/cockroachdb/cockroach/pkg/testutils.SucceedsWithin({0x8bc1cb8, 0xc00fff8000}, 0x8b127e0, 0x5a90c20)
        	/go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:60 +0x7d
        github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon({0x8bc1cb8, 0xc00fff8000}, 0xc000a0347c)
        	/go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:41 +0x5b
        github.com/cockroachdb/cockroach/pkg/server_test.TestClusterVersionUpgrade(0xc00fff8000)
        	/go/src/github.com/cockroachdb/cockroach/pkg/server/version_cluster_test.go:273 +0xae6
        testing.tRunner(0xc00fff8000, 0x5e74010)
        	/usr/local/go/src/testing/testing.go:1259 +0x230
        created by testing.(*T).Run
        	/usr/local/go/src/testing/testing.go:1306 +0x727
    panic.go:661: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestClusterVersionUpgrade338561729
--- FAIL: TestClusterVersionUpgrade (254.79s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- GOFLAGS=-json
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestClusterVersionUpgrade.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

server.TestClusterVersionUpgrade [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4289552&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4289552&tab=artifacts#/) on master @ [5c0d5132ade6d2e30a9d414163851272750d8afa](https://github.com/cockroachdb/cockroach/commits/5c0d5132ade6d2e30a9d414163851272750d8afa):


```
=== RUN   TestClusterVersionUpgrade
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestClusterVersionUpgrade3877796030
    test_log_scope.go:80: use -show-logs to present logs inline
    version_cluster_test.go:273: condition failed to evaluate within 45s: 0: v21.2-52 active=true (wanted false)
        goroutine 39611831 [running]:
        runtime/debug.Stack()
        	/usr/local/go/src/runtime/debug/stack.go:24 +0x65
        github.com/cockroachdb/cockroach/pkg/testutils.SucceedsWithin({0x509a198, 0xc00fe5a9c0}, 0x0, 0x3d3e7a0)
        	/go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:60 +0x5f
        github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon({0x509a198, 0xc00fe5a9c0}, 0xc00ada75e8)
        	/go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:41 +0x4a
        github.com/cockroachdb/cockroach/pkg/server_test.TestClusterVersionUpgrade(0xc00fe5a9c0)
        	/go/src/github.com/cockroachdb/cockroach/pkg/server/version_cluster_test.go:273 +0x8f6
        testing.tRunner(0xc00fe5a9c0, 0x41268e0)
        	/usr/local/go/src/testing/testing.go:1259 +0x102
        created by testing.(*T).Run
        	/usr/local/go/src/testing/testing.go:1306 +0x35a
    panic.go:642: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestClusterVersionUpgrade3877796030
--- FAIL: TestClusterVersionUpgrade (60.68s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- TAGS=deadlock

- GOFLAGS=-parallel=4
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestClusterVersionUpgrade.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

server.TestClusterVersionUpgrade [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4410859&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4410859&tab=artifacts#/) on master @ [89198dd89a0248a6344296138431c4f84dd09913](https://github.com/cockroachdb/cockroach/commits/89198dd89a0248a6344296138431c4f84dd09913):


```
=== RUN   TestClusterVersionUpgrade
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestClusterVersionUpgrade2998515523
    test_log_scope.go:80: use -show-logs to present logs inline
    version_cluster_test.go:271: condition failed to evaluate within 3m45s: 0: v21.2-68 active=true (wanted false)
        goroutine 574656 [running]:
        runtime/debug.Stack()
        	/usr/local/go/src/runtime/debug/stack.go:24 +0x72
        github.com/cockroachdb/cockroach/pkg/testutils.SucceedsWithin({0x8d8cb98, 0xc0022a2000}, 0x8cdc4e0, 0x5b3f540)
        	/go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:60 +0x7d
        github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon({0x8d8cb98, 0xc0022a2000}, 0xc00555364c)
        	/go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:41 +0x5b
        github.com/cockroachdb/cockroach/pkg/server_test.TestClusterVersionUpgrade(0xc0022a2000)
        	/go/src/github.com/cockroachdb/cockroach/pkg/server/version_cluster_test.go:271 +0xaa7
        testing.tRunner(0xc0022a2000, 0x5f2f420)
        	/usr/local/go/src/testing/testing.go:1259 +0x230
        created by testing.(*T).Run
        	/usr/local/go/src/testing/testing.go:1306 +0x727
    panic.go:661: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestClusterVersionUpgrade2998515523
--- FAIL: TestClusterVersionUpgrade (243.91s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- GOFLAGS=-json
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestClusterVersionUpgrade.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

server.TestClusterVersionUpgrade [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4437739&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4437739&tab=artifacts#/) on master @ [c518952e739509c8cfe8e3a93c2b6ffb3002b866](https://github.com/cockroachdb/cockroach/commits/c518952e739509c8cfe8e3a93c2b6ffb3002b866):


```
=== RUN   TestClusterVersionUpgrade
    test_log_scope.go:79: test logs captured to: /artifacts/tmp/_tmp/ab5462d2cb3989e2a33018db970f23de/logTestClusterVersionUpgrade3840330042
    test_log_scope.go:80: use -show-logs to present logs inline
    version_cluster_test.go:271: condition failed to evaluate within 45s: 0: v21.2-80 active=true (wanted false)
        goroutine 43155277 [running]:
        runtime/debug.Stack()
        	GOROOT/src/runtime/debug/stack.go:24 +0x65
        github.com/cockroachdb/cockroach/pkg/testutils.SucceedsWithin({0x5338df8, 0xc003cc2ea0}, 0x0, 0x3f57fc0)
        	github.com/cockroachdb/cockroach/pkg/testutils/soon.go:60 +0x5f
        github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon({0x5338df8, 0xc003cc2ea0}, 0xc0130cb2b0)
        	github.com/cockroachdb/cockroach/pkg/testutils/soon.go:41 +0x4a
        github.com/cockroachdb/cockroach/pkg/server_test.TestClusterVersionUpgrade(0xc003cc2ea0)
        	github.com/cockroachdb/cockroach/pkg/server_test/pkg/server/version_cluster_test.go:271 +0x896
        testing.tRunner(0xc003cc2ea0, 0x43699d8)
        	GOROOT/src/testing/testing.go:1259 +0x102
        created by testing.(*T).Run
        	GOROOT/src/testing/testing.go:1306 +0x35a
    panic.go:642: -- test log scope end --
--- FAIL: TestClusterVersionUpgrade (51.32s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- TAGS=bazel,gss,deadlock
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestClusterVersionUpgrade.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

server.TestClusterVersionUpgrade [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4468546&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4468546&tab=artifacts#/) on master @ [eee8ef080a95bc59eb3ddc2bebf623eb78e78db9](https://github.com/cockroachdb/cockroach/commits/eee8ef080a95bc59eb3ddc2bebf623eb78e78db9):


```
=== RUN   TestClusterVersionUpgrade
    test_log_scope.go:79: test logs captured to: /artifacts/tmp/_tmp/ab5462d2cb3989e2a33018db970f23de/logTestClusterVersionUpgrade2843081811
    test_log_scope.go:80: use -show-logs to present logs inline
    version_cluster_test.go:271: condition failed to evaluate within 45s: 0: v21.2-82 active=true (wanted false)
        goroutine 43244189 [running]:
        runtime/debug.Stack()
        	GOROOT/src/runtime/debug/stack.go:24 +0x65
        github.com/cockroachdb/cockroach/pkg/testutils.SucceedsWithin({0x5378918, 0xc002b2c680}, 0x0, 0x3f7e7a0)
        	github.com/cockroachdb/cockroach/pkg/testutils/soon.go:60 +0x5f
        github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon({0x5378918, 0xc002b2c680}, 0xc0066247f0)
        	github.com/cockroachdb/cockroach/pkg/testutils/soon.go:41 +0x4a
        github.com/cockroachdb/cockroach/pkg/server_test.TestClusterVersionUpgrade(0xc002b2c680)
        	github.com/cockroachdb/cockroach/pkg/server_test/pkg/server/version_cluster_test.go:271 +0x896
        testing.tRunner(0xc002b2c680, 0x4392230)
        	GOROOT/src/testing/testing.go:1259 +0x102
        created by testing.(*T).Run
        	GOROOT/src/testing/testing.go:1306 +0x35a
    panic.go:642: -- test log scope end --
--- FAIL: TestClusterVersionUpgrade (52.55s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- TAGS=bazel,gss,deadlock
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestClusterVersionUpgrade.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

server.TestClusterVersionUpgrade [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4550905&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4550905&tab=artifacts#/) on master @ [0069e980d756ca580167eb321ac6b348b35313f0](https://github.com/cockroachdb/cockroach/commits/0069e980d756ca580167eb321ac6b348b35313f0):


```
=== RUN   TestClusterVersionUpgrade
    test_log_scope.go:79: test logs captured to: /artifacts/tmp/_tmp/ab5462d2cb3989e2a33018db970f23de/logTestClusterVersionUpgrade2328019947
    test_log_scope.go:80: use -show-logs to present logs inline
    version_cluster_test.go:271: condition failed to evaluate within 45s: 0: v21.2-90 active=true (wanted false)
        goroutine 44618032 [running]:
        runtime/debug.Stack()
        	GOROOT/src/runtime/debug/stack.go:24 +0x65
        github.com/cockroachdb/cockroach/pkg/testutils.SucceedsWithin({0x5409338, 0xc012d78820}, 0x0, 0x3fe4200)
        	github.com/cockroachdb/cockroach/pkg/testutils/soon.go:60 +0x5f
        github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon({0x5409338, 0xc012d78820}, 0xc009523a20)
        	github.com/cockroachdb/cockroach/pkg/testutils/soon.go:41 +0x4a
        github.com/cockroachdb/cockroach/pkg/server_test.TestClusterVersionUpgrade(0xc012d78820)
        	github.com/cockroachdb/cockroach/pkg/server_test/pkg/server/version_cluster_test.go:271 +0x896
        testing.tRunner(0xc012d78820, 0x4401568)
        	GOROOT/src/testing/testing.go:1259 +0x102
        created by testing.(*T).Run
        	GOROOT/src/testing/testing.go:1306 +0x35a
    panic.go:642: -- test log scope end --
--- FAIL: TestClusterVersionUpgrade (52.36s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- TAGS=bazel,gss,deadlock
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestClusterVersionUpgrade.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

server.TestClusterVersionUpgrade [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4562624&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4562624&tab=artifacts#/) on master @ [9cdc505b79cde668e10d7a62109f4df313eda754](https://github.com/cockroachdb/cockroach/commits/9cdc505b79cde668e10d7a62109f4df313eda754):


```
=== RUN   TestClusterVersionUpgrade
    test_log_scope.go:79: test logs captured to: /artifacts/tmp/_tmp/ab5462d2cb3989e2a33018db970f23de/logTestClusterVersionUpgrade3504505469
    test_log_scope.go:80: use -show-logs to present logs inline
    version_cluster_test.go:271: condition failed to evaluate within 45s: 0: v21.2-92 active=true (wanted false)
        goroutine 44277160 [running]:
        runtime/debug.Stack()
        	GOROOT/src/runtime/debug/stack.go:24 +0x65
        github.com/cockroachdb/cockroach/pkg/testutils.SucceedsWithin({0x540f598, 0xc0043b21a0}, 0x0, 0x3fe8a60)
        	github.com/cockroachdb/cockroach/pkg/testutils/soon.go:60 +0x5f
        github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon({0x540f598, 0xc0043b21a0}, 0xc000d38b10)
        	github.com/cockroachdb/cockroach/pkg/testutils/soon.go:41 +0x4a
        github.com/cockroachdb/cockroach/pkg/server_test.TestClusterVersionUpgrade(0xc0043b21a0)
        	github.com/cockroachdb/cockroach/pkg/server_test/pkg/server/version_cluster_test.go:271 +0x896
        testing.tRunner(0xc0043b21a0, 0x4406260)
        	GOROOT/src/testing/testing.go:1259 +0x102
        created by testing.(*T).Run
        	GOROOT/src/testing/testing.go:1306 +0x35a
    panic.go:642: -- test log scope end --
--- FAIL: TestClusterVersionUpgrade (52.32s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- TAGS=bazel,gss,deadlock
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestClusterVersionUpgrade.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

server.TestClusterVersionUpgrade [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4566850&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4566850&tab=artifacts#/) on master @ [231fd420c61e39b6ad08c982c496d82cf1910bc5](https://github.com/cockroachdb/cockroach/commits/231fd420c61e39b6ad08c982c496d82cf1910bc5):


```
=== RUN   TestClusterVersionUpgrade
    test_log_scope.go:79: test logs captured to: /artifacts/tmp/_tmp/ab5462d2cb3989e2a33018db970f23de/logTestClusterVersionUpgrade1874473190
    test_log_scope.go:80: use -show-logs to present logs inline
    version_cluster_test.go:271: condition failed to evaluate within 45s: 0: v21.2-92 active=true (wanted false)
        goroutine 43497429 [running]:
        runtime/debug.Stack()
        	GOROOT/src/runtime/debug/stack.go:24 +0x65
        github.com/cockroachdb/cockroach/pkg/testutils.SucceedsWithin({0x540f618, 0xc002c10000}, 0x0, 0x3fe8ae0)
        	github.com/cockroachdb/cockroach/pkg/testutils/soon.go:60 +0x5f
        github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon({0x540f618, 0xc002c10000}, 0xc025f38400)
        	github.com/cockroachdb/cockroach/pkg/testutils/soon.go:41 +0x4a
        github.com/cockroachdb/cockroach/pkg/server_test.TestClusterVersionUpgrade(0xc002c10000)
        	github.com/cockroachdb/cockroach/pkg/server_test/pkg/server/version_cluster_test.go:271 +0x896
        testing.tRunner(0xc002c10000, 0x44062e0)
        	GOROOT/src/testing/testing.go:1259 +0x102
        created by testing.(*T).Run
        	GOROOT/src/testing/testing.go:1306 +0x35a
    panic.go:642: -- test log scope end --
--- FAIL: TestClusterVersionUpgrade (52.56s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- TAGS=bazel,gss,deadlock
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestClusterVersionUpgrade.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/ccl/kvccl/kvtenantccl/BUILD.bazel
+++ b/pkg/ccl/kvccl/kvtenantccl/BUILD.bazel
@@ -69,6 +69,7 @@ go_test(
         "//pkg/settings",
         "//pkg/settings/cluster",
         "//pkg/sql",
+        "//pkg/sql/sqlliveness/slinstance",
         "//pkg/testutils",
         "//pkg/testutils/serverutils",
         "//pkg/testutils/sqlutils",

--- a/pkg/ccl/kvccl/kvtenantccl/tenant_upgrade_test.go
+++ b/pkg/ccl/kvccl/kvtenantccl/tenant_upgrade_test.go
@@ -13,6 +13,7 @@ import (
 	gosql "database/sql"
 	"net/url"
 	"testing"
+	"time"
 
 	"github.com/cockroachdb/cockroach/pkg/base"
 	"github.com/cockroachdb/cockroach/pkg/clusterversion"
@@ -23,6 +24,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/security"
 	"github.com/cockroachdb/cockroach/pkg/server"
 	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
+	"github.com/cockroachdb/cockroach/pkg/sql/sqlliveness/slinstance"
 	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/testcluster"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
@@ -171,18 +173,13 @@ func TestTenantUpgrade(t *testing.T) {
 // equal the TestingBinaryMinSupportedVersion to avoid rot in tests using this
 // (as we retire old versions).
 func v0v1v2() (roachpb.Version, roachpb.Version, roachpb.Version) {
-	v1 := clusterversion.TestingBinaryVersion
 	v0 := clusterversion.TestingBinaryMinSupportedVersion
+	v1 := clusterversion.TestingBinaryVersion
 	v2 := clusterversion.TestingBinaryVersion
-	if v0.Minor > 0 {
-		v0.Minor--
+	if v1.Internal > 2 {
+		v1.Internal -= 2
 	} else {
-		v0.Major--
-	}
-	if v1.Minor > 0 {
-		v1.Minor--
-	} else {
-		v1.Major--
+		v2.Internal += 2
 	}
 	return v0, v1, v2
 }
@@ -237,6 +234,8 @@ func TestTenantUpgradeFailure(t *testing.T) {
 			v0,
 			false, // initializeVersion
 		)
+		slinstance.DefaultTTL.Override(ctx, &settings.SV, 3*time.Second)
+		slinstance.DefaultHeartBeat.Override(ctx, &settings.SV, 500*time.Millisecond)
 		v2onMigrationStopper := stop.NewStopper()
 		// Initialize the version to the minimum it could be.
 		require.NoError(t, clusterversion.Initialize(ctx,
@@ -246,6 +245,7 @@ func TestTenantUpgradeFailure(t *testing.T) {
 			TenantID: roachpb.MakeTenantID(id),
 			Existing: existing,
 			TestingKnobs: base.TestingKnobs{
+				JobsTestingKnobs: jobs.NewTestingKnobsWithShortIntervals(),
 				MigrationManager: &migration.TestingKnobs{
 					ListBetweenOverride: func(from, to clusterversion.ClusterVersion) []clusterversion.ClusterVersion {
 						return []clusterversion.ClusterVersion{{Version: v1}, {Version: v2}}

--- a/pkg/server/version_cluster_test.go
+++ b/pkg/server/version_cluster_test.go
@@ -28,7 +28,6 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/server"
 	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
 	"github.com/cockroachdb/cockroach/pkg/testutils"
-	"github.com/cockroachdb/cockroach/pkg/testutils/skip"
 	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/testcluster"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
@@ -208,7 +207,7 @@ func TestClusterVersionPersistedOnJoin(t *testing.T) {
 func TestClusterVersionUpgrade(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
-	skip.WithIssue(t, 74599, "flaky test")
+
 	ctx := context.Background()
 
 	var newVersion = clusterversion.TestingBinaryVersion

--- a/pkg/sql/set_cluster_setting.go
+++ b/pkg/sql/set_cluster_setting.go
@@ -13,6 +13,7 @@ package sql
 import (
 	"bytes"
 	"context"
+	"encoding/base64"
 	"fmt"
 	"strconv"
 	"strings"
@@ -45,6 +46,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/retry"
 	"github.com/cockroachdb/errors"
 	"github.com/cockroachdb/errors/hintdetail"
+	"github.com/cockroachdb/redact"
 )
 
 // setClusterSettingNode represents a SET CLUSTER SETTING statement.
@@ -454,7 +456,7 @@ func setVersionSetting(
 	}
 
 	// Updates the version inside the system.settings table.
-	// If we are already at the target version, then this
+	// If we are already at or above the target version, then this
 	// function is idempotent.
 	updateVersionSystemSetting := func(ctx context.Context, version clusterversion.ClusterVersion) error {
 		rawValue, err := protoutil.Marshal(&version)
@@ -468,24 +470,35 @@ func setVersionSetting(
 				sessiondata.InternalExecutorOverride{User: security.RootUserName()},
 				"SELECT value FROM system.settings WHERE name = $1", name,
 			)
-			versionIsDifferent := true
+			if err != nil {
+				return err
+			}
 			if len(datums) > 0 {
 				dStr, ok := datums[0].(*tree.DString)
 				if !ok {
-					return errors.Errorf("existing version value is not a string, got %T", datums[0])
+					return errors.AssertionFailedf("existing version value is not a string, got %T", datums[0])
 				}
 				oldRawValue := []byte(string(*dStr))
-				versionIsDifferent = !bytes.Equal(oldRawValue, rawValue)
+				if bytes.Equal(oldRawValue, rawValue) {
+					return nil
+				}
+				var oldValue clusterversion.ClusterVersion
+				// If there is a pre-existing value, we ought to be able to decode it.
+				if err := protoutil.Unmarshal(oldRawValue, &oldValue); err != nil {
+					return errors.NewAssertionErrorWithWrappedErrf(err, "decoding previous version %s",
+						redact.SafeString(base64.StdEncoding.EncodeToString(oldRawValue)))
+				}
+				if !oldValue.Less(version.Version) {
+					return nil
+				}
 			}
-			// Only if the version has changed alter the setting.
-			if versionIsDifferent {
-				_, err = execCfg.InternalExecutor.ExecEx(
-					ctx, "update-setting", txn,
-					sessiondata.InternalExecutorOverride{User: security.RootUserName()},
-					`UPSERT INTO system.settings (name, value, "lastUpdated", "valueType") VALUES ($1, $2, now(), $3)`,
-					name, string(rawValue), setting.Typ(),
-				)
-			}
+			// Only if the version has increased, alter the setting.
+			_, err = execCfg.InternalExecutor.ExecEx(
+				ctx, "update-setting", txn,
+				sessiondata.InternalExecutorOverride{User: security.RootUserName()},
+				`UPSERT INTO system.settings (name, value, "lastUpdated", "valueType") VALUES ($1, $2, now(), $3)`,
+				name, string(rawValue), setting.Typ(),
+			)
 			return err
 		})
 	}

